### ISSUE TITLE
Incremental compile processor should consider externally declared macros

### ISSUE THREAD

**Issue by fletcher86** [gradle/gradle-native#882]
### Expected Behavior
Given a simple build sample.  Having already done a gradle build previously, gradle does a full build of each and every source file again after a simple modification of a single file.  This scenario occurs when header file contains a macro include and is an include within the source files.  The expected behavior is that gradle should do an incremental build for only the file that has changed.

Here is the structure of the build sample.  Or you can clone it directly from 
`git clone https://github.com/fletcher86/macro-includes`

`src/main/cpp/`

main.cpp
`#include "main.h"`
`int main() {`
`    printf("Hello, World!\n");`
`    return 0;`
`}`

main1.cpp
`#include "main.h"`
`int function1() {`
`    printf("Call F1!\n");`
`    return 0;`
`}`

main2.cpp
`#include "main.h"`
`int function2() {`
`    printf("Call F2\n");`
`    return 0;`
`}`

`src/main/headers`
main.h
`#include STDIO_H`

build.gradle
`plugins {`
`  id 'cpp-application'`
`}` 
`application {`
`    source.from project.fileTree(dir: 'src/main/cpp', include: '*.cpp') `
`}`
`tasks.withType(CppCompile) { `
`    macros.put('STDIO_H','"stdio.h"')`
`}`

### Current Behavior
Assuming you have already done a full build previously, if you change a single character (touch) in main.cpp, save your file, then run `gradle build` again, gradle while compile each and every file, main.ccp, main1.cpp, and main2.cpp.

### Context

We are trying to port a project that we do not own the code for to gradle with thousands of source files.  This is very time consuming to do a full build when only one file has changed and impacting our ability to optimize the build.  For this particular example, doing a full build isn't a big deal, however, for larger projects, this appears to be a big penalty.
### Steps to Reproduce (for bugs)

`git clone https://github.com/fletcher86/macro-includes`
`cd macro-includes`
`./gradlew build`
`touch src/main/cpp/main.cpp`
`./gradlew.bat build`

Now examine the 'output.txt' to see that every file was compiled when it was expected that only main.cpp should have been compiled.

`See file:///E:/macro-includes/build/tmp/compileDebugCpp/output.txt for all output for compileDebugCpp.`

`compiling main1.cpp successful.`
`main1.cpp`

`compiling main.cpp successful.`
`main.cpp`

`compiling main2.cpp successful.`
`main2.cpp`

`Finished compileDebugCpp, see full log file:///E:/macro-includes/build/tmp/compileDebugCpp/output.txt.`

### Your Environment

 * Build scan URL: 
https://gradle.com/s/6jzl2jxdqyrhi

**Comment by lacasseio** [comment]
Thanks for reporting on this use case. We are aware of similar issues around macro includes and needs to prioritize according to real life projects. Gradle prioritize been correct in all cases. If Gradle can't follow the headers during discovery, we revert back to the most correct behavior which is rebuilding the file. In your case, all three compilation units are similar in design and touch a part of the analysis that we don't support yet. Gradle revert to compiling all three files.

I can see that the `STDIO_H` macro is defined outside of the source and it's never undefined. If this is always the case, we can optimize for this and keep the incrementally. Can you confirm that the macros used for macro includes are defined outside of the source and they are never undefined within the source code?

**Comment by amz-shahji** [comment]
Option to revert back to full rebuild works for projects with 4 files not with 4K files and 2 hour build times. Yes, it's a "safe solution" but not really a solution. Firstly, making this more obvious without digging deep into logs would be worth it.

As for the use case, the sample provided is just an example built to demonstrate the problem. Real world problems are more demanding. For example, boost uses this in a whole lot of files. It doesn't even have a build step because its all just headers. This dependency comes in transitively into many projects because our foundation project is built on it. We have projects where the macro is provided on command line (like in the example) and other examples where the macro expansion is a result of various conditions evaluated at preprocessor time (platform, build environment, architecture, and at times even compiler supported features like clang __has_feature).

Bottom-line the source of the macro and how it get expanded souldn't be part of the solution to make this work.

**Comment by amz-shahji** [comment]
Now as for the solution, it more likely going to be compiler dependent. For example, in case of vc/cl, use of option /showIncludes [https://msdn.microsoft.com/en-us/library/hdkef6tk.aspx](https://msdn.microsoft.com/en-us/library/hdkef6tk.aspx) can provide a list of headers that any compilation unit is dependent on. When making full builds (i.e. non-incremental builds), the output can be scanned and cached for recurring incremental builds.

For gcc, there are similar options available [https://stackoverflow.com/questions/4479049/show-include-equivalent-option-in-g](https://stackoverflow.com/questions/4479049/show-include-equivalent-option-in-g)

And yes, this solution takes the header scanning off of the hands of Gradle.

**Comment by lacasseio** [comment]
Thanks @amz-shahji for the additional information and we are aware of the various limitation of the current approach where Gradle parse himself the headers versus handing off the scanning to the compiler as you describe. Don't get me wrong, I'm a big advocate around using the compiler instead. However, after long discussion internally, there's a lot of advantage, in the long run, to have Gradle parse the headers and we haven't had a compelling argument for changing our strategy. The idea is to make to common case fast and the exception fast enough. This is why we want to have a real-life use case before optimizing for them. To be frank everything in native preprocessing is possible from simple feature flag to full encryption through clever macro manipulation. We are sorting out what at the most likely cases and optimizing those first.

In your case, it seems you are hitting a strange issue where Gradle is able to handle macro include such as:
```
#define STDIO_H "stdio.h"  // in upstream source files
#include "main.h"
int main() {
printf("Hello, World!\n");
return 0;
}
```
as well as
```
#define STDIO_H "stdio.h"  // side-by-side with the macro include
#include STDIO_H
```
However, we have a gap when the macro is defined through the `CppCompile#macro` method. Working on solving this should show a significant gain in performance as recompilation won't be retriggered for all files transitively including a header whose macro is defined outside the build. I will rename this specific issue to handling this use case.

We have other clients having lots of success and acceptable performance using Boot libraries and all it's macro using. I would fix this issue and profile the build before moving onto a different implementation of the header analysis.

**Comment by amz-shahji** [comment]
Fair enough, if you see value in pursuing specific use cases. I am not familiar with all the intricacies of handling the issue so I will reserve my comments to providing use cases. Along that stream, here are the current use case related to using boost (and lua, which is covered by the following scenario).

Cannot locate header file for '#include AUX778076_INCLUDE_STRING' in source file 'include_preprocessed.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_1' in source file 'forward1.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_1' in source file 'reverse1.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_2' in source file 'forward2.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_2' in source file 'reverse2.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_3' in source file 'forward3.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_3' in source file 'reverse3.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_4' in source file 'forward4.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_4' in source file 'reverse4.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_5' in source file 'forward5.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_FILENAME_5' in source file 'reverse5.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'advance_backward.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'advance_forward.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'apply.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'apply_fwd.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'apply_wrap.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'bind.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'bind_fwd.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'fold_impl_body.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'full_lambda.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'lambda_no_ctps.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'numeric_op.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'placeholders.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'quote.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'reverse_fold_impl_body.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'sequence_wrapper.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'substitute.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_ITERATE()' in source file 'template_arity.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_STRINGIZE(boost / mpl / aux_ / preprocessed / AUX778076_PREPROCESSED_HEADER)' in source file 'include_preprocessed.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_STRINGIZE(boost / mpl / list / AUX778076_LIST_HEADER)' in source file 'list.hpp'. Assuming changed.
Cannot locate header file for '#include BOOST_PP_STRINGIZE(boost / mpl / vector / AUX778076_VECTOR_HEADER)' in source file 'vector.hpp'. Assuming changed.

Specific headers can be found in boost source code. I see these fall in few different broad categories -

1. Non-parameterized macro - which seems to be handled as long as the macro is visible at the time of parsing the header. Value coming from build script is not handled.
2. Parameterized macro - I don't think this is handled in gradle.
3. String concatenation - Easiest of all, if individual tokens are immediately visible. If not, this will fall in the same category as 1. in this list.

Any guidance on fixing these in gradle source will help tremendously to unblock our progress.

**Comment by adammurdoch** [comment]
The impact of this issue isn't as great as the description suggests, depending on how your source is arranged. Gradle will only recompile a source file if at least one of these is true:

1. The content of source file itself has changed. Using `touch src/main/cpp/main.cpp` won't trigger this case, as this changes the timestamp of the source file but not the _content_.
2. The content of a header file included by that source file has changed, whether included directly or indirectly. Again, using `touch something.hpp` won't trigger this case as this does not change the content of the header file.
3. The source file or one of the headers it includes contains an unresolveable macro, such as the example in the description of this issue has _and_ the content of any header file has changed.

The "steps to reproduce" in the description are broken. If you touch a source file and run `gradlew build` then Gradle does nothing because no files have changed. If you then look at `output.txt`, then you're looking at the `output.txt` from the _previous_ invocation of Gradle (that is, the invocation that actually built everything).

A better experiment would be to append an empty line (or a comment) to the end of the source file. If you do this for a source file, you see that Gradle only recompiles that source file and nothing else. It's only when every source file contains an unresolveable macro (like in the example) _and_ you change a header file that Gradle will recompile everything.

This isn't to say we shouldn't fix the issue, only that the impact isn't actually as great as the description suggests, for many common kinds of source code changes (eg I change a function body and rebuild).

**Comment by adammurdoch** [comment]
This is fixed via https://github.com/gradle/gradle/pull/7375. The fix didn't quite make it in time for Gradle 5.0 but will be included in Gradle 5.1.

**Comment by lacasseio** [comment]
@amz-shahji @fletcher86 Would you mind [trying the latest nightly](https://gradle.org/nightly/) to see how our fix improve your situation?

**Comment by amz-shahji** [comment]
Fix confirmed.

From 4.10.2, I see the following message in the log

`After parsing the source files, Gradle cannot calculate the exact set of include files for ':compileDebugCpp'. Every file in the include search path will be considered a header dependency.`

Which is no more in the nightly build and it does seem to be doing the right thing.

Appreciate the prompt attention to this issue.
